src/ucp/core/ucp_worker.c

```diff
@@ -2453,44 +2454,85 @@ static int ucp_worker_keepalive_is_enabled(ucp_worker_h worker)
+                                                ucp_ep_ext_gen_t, ep_list);
+
+    return ucp_ep_from_ext_gen(ep_ext);
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_worker_keepalive_next_ep(ucp_worker_h worker)
+{
+    ucp_ep_h ep;
+
+    worker->keepalive.iter = worker->keepalive.iter->next;
+    if (worker->keepalive.iter == &worker->all_eps) {
+        /* if next list item points to all_eps then step one more time */
+        worker->keepalive.iter = worker->keepalive.iter->next;
+    }
+
+    ucs_assert(worker->keepalive.iter != &worker->all_eps);
+    ep                         = ucp_worker_keepalive_current_ep(worker);
+    worker->keepalive.lane_map = ((ep->cfg_index != UCP_WORKER_CFG_INDEX_NULL) &&
+                                  !(ep->flags & UCP_EP_FLAG_FAILED)) ?
+                                 ucp_ep_config(ep)->key.ep_check_map : 0;
+}
+
 static unsigned ucp_worker_keepalive_progress(void *arg)
 {
     ucp_worker_h worker = (ucp_worker_h)arg;
     ucs_time_t now      = ucs_get_time();
     unsigned ep_keepalive_count;
-    ucp_ep_ext_gen_t *ep_ext;
     ucs_list_link_t *iter_begin;
+    ucp_ep_h ep;
 
     if (ucs_likely((now - worker->keepalive.last_round) <
                    worker->context->config.ext.keepalive_timeout)) {
         return 0;
     }
 
-    worker->keepalive.last_round = now;
-    ep_keepalive_count           = 0;
-    iter_begin                   = worker->keepalive.iter;
+    ep_keepalive_count = 0;
+    if (ucs_unlikely(ucs_list_is_empty(&worker->all_eps))) {
+        ucs_assert(worker->keepalive.iter == &worker->all_eps);
+        goto out;
+    }
+
+    if (ucs_unlikely(worker->keepalive.iter == &worker->all_eps)) {
+        ucp_worker_keepalive_next_ep(worker);
+    }
+
+    iter_begin = worker->keepalive.iter;
     /* use own loop for elements because standard for_each skips
      * head element */
     /* TODO: use more optimal algo to enumerate EPs to keepalive
      * (linked list) */
     do {
-        ep_ext = ucs_list_next(worker->keepalive.iter,
-                               ucp_ep_ext_gen_t, ep_list);
-        worker->keepalive.iter = &ep_ext->ep_list;
-        if (worker->keepalive.iter == &worker->all_eps) {
-            continue;
+        ep = ucp_worker_keepalive_current_ep(worker);
+        ucp_ep_do_keepalive(ep, &worker->keepalive.lane_map);
+        if (worker->keepalive.lane_map != 0) {
+            /* in case if EP has no resources to send keepalive message
+             * then just return without update of last_round timestamp,
+             * on next progress iteration we will continue from this point */
+            goto out_no_resources;
         }
 
-        ep_keepalive_count += ucp_ep_do_keepalive(ucp_ep_from_ext_gen(ep_ext));
-    } while ((iter_begin != &ep_ext->ep_list) &&
+        ep_keepalive_count++;
+        ucp_worker_keepalive_next_ep(worker);
+    } while ((iter_begin != worker->keepalive.iter) &&
```

Comment:
shouldn't we stop when reached &worker->all_eps?
after this PR we can exit in middle of the round
also, we need to remember count from last round: if 1st round we sent keepalive_num_eps,  and then found "NO_RESOURCE", we will try to send again immediately and can send keepalive_num_eps more. So as a result we will send keepalive_num_eps*2 in same time unit.

Reply:
ucp_worker_keepalive_next_ep doesn't stop at head element...
I added counter into worker->keepaline structure